### PR TITLE
Mkerns/caseflow 1946 create vha program office org queue

### PR DESCRIPTION
**Pull request by mkerns355**
Resolves [#1946](https://vajira.max.gov/browse/CASEFLOW-1946)

### Description
Added Custom Program Office Queues w/ custom table columns. 

### Acceptance Criteria
- [x] Code compiles correctly
- [x] Create queues for the following Program Office organizations:
- [x] Community Care - Payment Operations Management
- [x] Community Care - Veteran and Family Members Program
- [x] Member Services - Health Eligibility Center
- [x] Member Services - Beneficiary Travel
- [x] Prosthetics
- [x] Caregiver Appeals

### **Assigned**
- [x] Cases received from CAMO and yet to be worked
- [x] Cases returned by VISN that need to be reassigned

### **In progress**
- [x] Cases marked in progress by Program Office users
- [ ] Cases assigned to a VISN/VAMC

### **Ready for review**
- [x] Cases received back from the VISN that have a completed documentation check

### **On hold**
- [x] Cases on hold at the Program Office level
- [x] Cases on hold at the VISN level

### **Completed**
- [x] Cases marked completed at the Program Office level

### **Updated columns:**

- [ ] Task Owner column will show either the assigned Program Office or if at a a VISN, the respective VISN/VAMC
- [ ] Added the VAMC column for additional labeling
- [ ] Last Action replaces "Days since last change"
- [ ] Board Intake replaces "Days Waiting"
- [ ] Board Intake column includes the amount of days and "ago"

### Testing Plan
1. Go to ...

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---

### Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

**Review by ajspotts — APPROVED** [review]
Commented out code is being left in for subsequent work on the missing additional columns (also at least one method that we're not using needs to be removed)

File: app/models/queue_tab.rb [hunk 1]
```diff
@@ -98,14 +99,38 @@ def on_hold_tasks
...
     Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).recently_completed
   end
 
+  # def without_child_tasks
+  #   Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).assigned.without_children
+  # end
+
+  def without_child_tasks
+    Task.where(children: nil)
+  end
```
Comment (yoomlam) [inline app/models/queue_tab.rb:120]: Is there a test for this? This will always return nil:
```ruby
puts Task.where(children: nil).to_sql
SELECT "tasks".* FROM "tasks" WHERE "tasks"."id" IS NULL
```

File: app/models/queue_tab.rb [hunk 2]
```diff
@@ -98,14 +99,38 @@ def on_hold_tasks
...
+    Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).cancelled
+  end
+
   def recently_completed_tasks
     Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).recently_completed
   end
 
+  # def without_child_tasks
+  #   Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).assigned.without_children
+  # end
```
Comment (yoomlam) [inline app/models/queue_tab.rb:116]: Why is this here?

File: app/models/queue_tabs/vha_program_office_assigned_tasks_tab.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def description
+    format(COPY::ORGANIZATIONAL_QUEUE_PAGE_ASSIGNED_TASKS_DESCRIPTION, assignee.name)
+  end
+
+  def parent_ids_with_completed_child_assess_documentation_task
+    assigned_task_children.where(type: :AssessDocumentationTask)
+      .completed
+      .pluck(:parent_id)
```
Comment (yoomlam) [inline app/models/queue_tabs/vha_program_office_assigned_tasks_tab.rb:21]: `.pluck(:parent_id).uniq`

File: app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+  def self.tab_name
+    Constants.QUEUE_CONFIG.READY_FOR_REVIEW_TASKS_TAB_NAME
+  end
+
+  def description
+    format(COPY::ORGANIZATIONAL_QUEUE_PAGE_READY_FOR_REVIEW_TASKS_DESCRIPTION, assignee.name)
+  end
+
+  def parents_with_completed_child_assess_documentation_task
```
Comment (yoomlam) [inline app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb:18]: Better method name: `parent_ids_with_completed_child_assess_documentation_task`

File: app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    assigned_task_children.where(type: :AssessDocumentationTask)
+      .completed
+      .pluck(:parent_id)
```
Comment (yoomlam) [inline app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb:21]: `.pluck(:parent_id).uniq`

File: app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def tasks
+    # byebug
```
Comment (yoomlam) [inline app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb:25]: Remove

File: app/models/serializers/work_queue/task_column_serializer.rb [hunk 7]
```diff
@@ -130,13 +130,24 @@ def self.serialize_attribute?(params, columns)
...
   end
 
+  # attribute :updated_at do |object, params|
+  #   columns = [Constants.QUEUE_CONFIG.COLUMNS.LAST_ACTION.name]
+
+  #   if serialize_attribute?(params, columns)
+  #     object.updated_at
+  #   end
+  # end
+
```
Comment (yoomlam) [inline app/models/serializers/work_queue/task_column_serializer.rb:150]: Why is this here?

File: app/models/task.rb [hunk 8]
```diff
@@ -329,6 +329,10 @@ def open_with_no_children?
     open? && children.empty?
   end
 
+  def active_with_no_children?
+    active? && children.empty?
+  end
+
```
Comment (yoomlam) [inline app/models/task.rb:335]: Add a test for this method. Where is this used?

File: client/app/queue/QueueTableBuilder.jsx [hunk 9]
```diff
@@ -12,20 +12,24 @@ import Link from '@department-of-veterans-affairs/caseflow-frontend-toolkit/comp
...
   assignedToColumn,
+  boardIntakeColumn,
   badgesColumn,
   completedToNameColumn,
   daysOnHoldColumn,
   daysWaitingColumn,
   detailsColumn,
   docketNumberColumn,
   documentIdColumn,
+  // lastActionColumn,
```
Comment (yoomlam) [inline client/app/queue/QueueTableBuilder.jsx:23]: Remove

File: client/app/queue/QueueTableBuilder.jsx [hunk 10]
```diff
@@ -113,6 +119,22 @@ class QueueTableBuilder extends React.PureComponent {
...
         filterOptions
       ),
+      [QUEUE_CONFIG.COLUMNS.BOARD_INTAKE.name]: boardIntakeColumn(
+        tasks,
+        filterOptions
+      ),
+      // [QUEUE_CONFIG.COLUMNS.LAST_ACTION.name]: lastActionColumn(
+      //   tasks,
+      //   filterOptions
+      // ),
```
Comment (yoomlam) [inline client/app/queue/QueueTableBuilder.jsx:129]: Remove

File: client/app/queue/components/TaskTableColumns.jsx [hunk 11]
```diff
@@ -109,6 +109,88 @@ export const detailsColumn = (tasks, requireDasRecord, userRole) => {
...
+// };
+
+// export const lastActionColumn = () => {
+//   return {
+//     header: COPY.CASE_LIST_TABLE_APPEAL_LAST_ACTION_COLUMN_TITLE,
+//     name: QUEUE_CONFIG.COLUMNS.LAST_ACTION.name,
+//     valueFunction: (task) => task.updatedAt,
+//     getSortValue: (task) => task.updatedAt
+//   };
+// };
```
Comment (yoomlam) [inline client/app/queue/components/TaskTableColumns.jsx:160]: remove

File: client/constants/QUEUE_CONFIG.json [hunk 12]
```diff
@@ -33,6 +43,11 @@
       "sorting_table": "tasks",
       "sorting_columns": ["assigned_at"]
     },
+    "LAST_ACTION": {
+      "name": "lastActionColumn",
+      "sorting_table": "tasks",
+      "sorting_columns": ["updated_at"]
+    },
```
Comment (yoomlam) [inline client/constants/QUEUE_CONFIG.json:50]: Is this used? If not, remove.

File: app/models/organizations/vha_program_office.rb [hunk 13]
```diff
@@ -4,4 +4,44 @@ class VhaProgramOffice < Organization
...
+    ::VhaProgramOfficeReadyForReviewTasksTab.new(assignee: self)
+  end
+
+  def on_hold_tasks_tab
+    ::VhaProgramOfficeOnHoldTasksTab.new(assignee: self)
+  end
+
+  def completed_tasks_tab
+    ::VhaProgramOfficeCompletedTasksTab.new(assignee: self)
+  end
```
Comment (yoomlam) [inline app/models/organizations/vha_program_office.rb:36]: Where are tests for these new tabs?

File: app/models/queue_tab.rb (hunk 1 above)
Comment (mkerns355) [reply app/models/queue_tab.rb:120]: Removing this in my follow up PR. Please disregard.

File: app/models/queue_tab.rb (hunk 2 above)
Comment (mkerns355) [reply app/models/queue_tab.rb:116]: Removing this in my follow up PR. Please disregard.

File: app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb (hunk 6 above)
Comment (mkerns355) [reply app/models/queue_tabs/vha_program_office_ready_for_review_tasks_tab.rb:25]: Removing this in my follow up PR. Please disregard.

File: app/models/serializers/work_queue/task_column_serializer.rb (hunk 7 above)
Comment (mkerns355) [reply app/models/serializers/work_queue/task_column_serializer.rb:150]: Will be using this in my follow up PR. Please disregard.

File: app/models/task.rb (hunk 8 above)
Comment (mkerns355) [reply app/models/task.rb:335]: Removing this in my follow up PR. Please disregard.

File: client/app/queue/QueueTableBuilder.jsx (hunk 9 above)
Comment (mkerns355) [reply client/app/queue/QueueTableBuilder.jsx:23]: Will be using this in my follow up PR. Please disregard.

File: client/app/queue/QueueTableBuilder.jsx (hunk 10 above)
Comment (mkerns355) [reply client/app/queue/QueueTableBuilder.jsx:129]: Will be using this in my follow up PR. Please disregard.

File: client/app/queue/components/TaskTableColumns.jsx (hunk 11 above)
Comment (mkerns355) [reply client/app/queue/components/TaskTableColumns.jsx:160]: Will be using some of this in my follow up PR. Please disregard.

File: client/constants/QUEUE_CONFIG.json (hunk 12 above)
Comment (mkerns355) [reply client/constants/QUEUE_CONFIG.json:50]: Will be using this in my follow up PR. Please disregard.

File: app/models/organizations/vha_program_office.rb (hunk 13 above)
Comment (mkerns355) [reply app/models/organizations/vha_program_office.rb:36]: Will be adding this in my follow up PR. Please disregard.
